### PR TITLE
Cryptographic primitives

### PR DESCRIPTION
**Pull request by abizjak**
## Purpose

Add cryptographic primitives to V1 smart contracts.

## Changes

Add
- ed25519 signature verification
- ecdsa over secp256k1 signature verification (bitcoin signatures)
- sha2, sha3 and keccak, all in 256 bit versions

## Checklist

- [ ] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: wasm-chain-integration/src/v1/types.rs [hunk 1]
```diff
@@ -416,6 +422,11 @@ impl Output for ImportFunc {
                 CommonFunc::StateEntrySize => 20,
                 CommonFunc::StateEntryResize => 21,
                 CommonFunc::WriteOutput => 22,
+                CommonFunc::VerifyEd25519 => 32,
```
Comment (Bargsteen) [inline wasm-chain-integration/src/v1/types.rs:425]: Why the gap? To add more state-related host functions near each other?

**Review by Bargsteen — APPROVED** [review]
LGTM :+1:

**Review by chrmatt — APPROVED** [review]
Looks good to me.

File: wasm-chain-integration/src/v1/types.rs (hunk 1 above)
Comment (abizjak) [reply wasm-chain-integration/src/v1/types.rs:430]: there is actually no gap. The Init/Receive functions are in the 22-31 range.

**Review by tschudid — COMMENTED** [review]
LGTM

File: wasm-chain-integration/benches/code/v1/host-functions.wat [hunk 2]
```diff
@@ -40,6 +40,13 @@
   ;; return a value
   (import "concordium" "write_output" (func $write_output (param $start i32) (param $length i32) (param $offset i32) (result i32)))
 
+  ;; cryptographic primitives
+  (import "concordium" "verify_ed25519_signature" (func $verify_ed25519_signature (param $public_key i32) (param $signature i32) (param $message i32) (param $message_len i32) (result i32)))
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/benches/code/v1/host-functions.wat:44]: The length is used by the execution engine to read that amount of data from the memory.

If the length is wrong then either signature verification will fail, or "illegal memory access" exception will be raised, depending on how wrong it is.

File: wasm-chain-integration/benches/code/v1/host-functions.wat [hunk 3]
```diff
@@ -40,6 +40,13 @@
...
+  (import "concordium" "verify_ecdsa_secp256k1_signature" (func $verify_ecdsa_secp256k1_signature (param $public_key i32) (param $signature i32) (param $message i32) (result i32)))
+  (import "concordium" "hash_sha2_256" (func $hash_sha2_256 (param $data i32) (param $data_len i32) (param $output i32)))
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/benches/code/v1/host-functions.wat:46]: The length is used by the execution engine to read that amount of data from the memory.

If the length is wrong then either signature verification will fail, or "illegal memory access" exception will be raised, depending on how wrong it is.

File: wasm-chain-integration/benches/v1-host-functions.rs [hunk 4]
```diff
@@ -343,6 +344,135 @@ pub fn criterion_benchmark(c: &mut Criterion) {
...
+
+    {
+        // n is the length of the data to be hashed
+        for n in [0u32, 10, 20, 50, 100, 1000, 10_000, 100_000] {
+            let name = "hostfn.hash_keccak_256";
+            let params = n.to_le_bytes().to_vec(); // length to hash
+            add_crypto_primitive_benchmark(name, params, Some(n));
+        }
+    }
+
```
Comment (abizjak) [inline wasm-chain-integration/benches/v1-host-functions.rs:475]: @tschudid These are the benchmarks.

They are done in the same way as all the other benchmarks to assign costs. They are run on the workstation we have here which is a representative mid-high end workstation.

**Review by tschudid — APPROVED** [review]

**Pull request merged**